lib/systemd/user-level-services.nix

```diff
@@ -0,0 +1,107 @@
+{ nixpkgs
+, system ? "x86_64-linux"
+, pkgs ? nixpkgs.legacyPackages.${system}, ... }:
+let
+  mkVM = moduleConfig: serviceConfig: nixpkgs.lib.nixosSystem {
+    inherit system;
+    modules = [ ({ ... }: {
+
+      imports = [ moduleConfig.module ];
+
+      services.${moduleConfig.nixOsServiceName} = {
+        enable = true;
+      } // serviceConfig;
+
+    }) ];
+  };
+
+in {
+  mkActivateScript = activateScriptName:
+      { module, nixOsServiceName, primaryService, auxiliaryServices}@moduleConfig:
+      { backupAction ? "", checkAction ? null, restoreAction ? ""}:
+      serviceConfig:
+    let
+      vm = mkVM moduleConfig serviceConfig;
+      allServices = [primaryService] ++ auxiliaryServices;
+    in
+    pkgs.writeShellScriptBin activateScriptName ''
+      set -euo pipefail
+      export XDG_RUNTIME_DIR="/run/user/$UID"
+      mkdir -v -p "$HOME/.config/systemd/user"
+      ${if backupAction != ""
+        then
+          ''
+            # Deploy may change DB schema via migration and we should be able
+            # to restore its state to a point before an update to successfully rollback
+            # failed deployment
+            echo Backing up the existing state
+            ${backupAction}
+          ''
+        else
+          ""
+      }
+
+      ${pkgs.lib.concatStringsSep "\n" (builtins.map (service:
+        let
+          serviceUnit = vm.config.systemd.units.${service};
+        in
+          ''
+            rm -v -f -- "$HOME/.config/systemd/user/${service}"
+            ln -v -s ${serviceUnit.unit}/${service} "$HOME/.config/systemd/user/${service}"
+          ''
+          +
+          (pkgs.lib.concatStringsSep "\n" (builtins.map (wantedByTarget:
+              ''
+                rm -v -f -- "$HOME/.config/systemd/user/${wantedByTarget}.wants/${service}"
+                mkdir -v -p "$HOME/.config/systemd/user/${wantedByTarget}.wants"
+                ln -v -s "$HOME/.config/systemd/user/${service}" "$HOME/.config/systemd/user/${wantedByTarget}.wants"
+              ''
+          ) serviceUnit.wantedBy))
+        ) allServices)}
+      systemctl --user daemon-reload
+
+      ${if restoreAction != ""
+        then
+          ''
+          rollback() {
+            exit_code="$?"
+            if [[ $exit_code -ne 0 ]]; then
+              echo Activation failed, restoring the previous working state
+              ${restoreAction}
+            fi
+            exit "$exit_code"
+          }
+          trap rollback EXIT
+          ''
+        else ""
```

Comment:
I think such interpolation blocks would look a little cleaner with `pkgs.lib.optionalString` instead of if statements

Reply:
applied